app/controllers/actions/RequiredUserAnswersForCorrectReturn.scala

```diff
@@ -0,0 +1,153 @@
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import models.correctReturn.{AddASmallProducer, RepaymentMethod}
+import models.requests.DataRequest
+import models.{CheckMode, LitresInBands, NormalMode, RetrievedSubscription, SdilReturn}
+import pages.{Page, QuestionPage}
+import pages.correctReturn._
+import play.api.libs.json.Reads
+import play.api.mvc.Result
+import play.api.mvc.Results.Redirect
+import utilities.{GenericLogger, UserTypeCheck}
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+
+class RequiredUserAnswersForCorrectReturn @Inject()(genericLogger: GenericLogger)(implicit val executionContext: ExecutionContext) extends ActionHelpers {
+
+  def requireData(page: Page)(action: => Future[Result])(implicit request: DataRequest[_]): Future[Result] = {
+    page match {
+      case CorrectReturnBaseCYAPage => checkYourAnswersRequiredData(action)
+      case CorrectReturnCheckChangesPage => checkChangesRequiredData(action)
+      case _ => action
+    }
+  }
+
+  private[controllers] def checkYourAnswersRequiredData(action: => Future[Result])(implicit request: DataRequest[_]): Future[Result] = {
+    val userAnswersMissing: List[CorrectReturnRequiredPage[_, _, _]] = returnMissingAnswers(mainRoute)
+    userAnswersMissing.headOption.map(_.pageRequired.asInstanceOf[Page]) match {
+      case Some(page) if List(PackAtBusinessAddressPage, PackagingSiteDetailsPage).contains(page) =>
+        genericLogger.logger.info(s"${request.userAnswers.id} now requires packaging sites")
+        Future.successful(Redirect(PackAtBusinessAddressPage.url(CheckMode)))
+      case Some(page) if page == SecondaryWarehouseDetailsPage =>
+        genericLogger.logger.info(s"${request.userAnswers.id} now has option to add warehouse")
+        Future.successful(Redirect(page.url(CheckMode)))
+      case Some(page) => genericLogger.logger.warn(s"${request.userAnswers.id} has hit correct return base CYA and is missing $userAnswersMissing")
+        Future.successful(Redirect(page.url(NormalMode)))
+      case None => action
+    }
+  }
+
+  private[controllers] def checkChangesRequiredData(action: => Future[Result])(implicit request: DataRequest[_]): Future[Result] = {
+    val userAnswersMissing: List[CorrectReturnRequiredPage[_, _, _]] = returnMissingAnswers(correctChangesJourney)
+    userAnswersMissing.headOption.map(_.pageRequired.asInstanceOf[Page]) match {
+      case Some(page) => genericLogger.logger.warn(s"${request.userAnswers.id} has hit check changes CYA and is missing $userAnswersMissing")
+        Future.successful(Redirect(page.url(NormalMode)))
+      case None => action
+    }
+  }
+
+  private[controllers] def returnMissingAnswers[A: ClassTag, B: ClassTag](list: List[CorrectReturnRequiredPage[_, _, _]])
+                                                                         (implicit request: DataRequest[_]): List[CorrectReturnRequiredPage[_, _, _]] = {
```

Comment:
If this were PreviousPage, rather than CorrectReturnPreviousPage, could perhaps pull out as a helper method. (If you thought it were neater - I think it may be, but not certain of it)

Reply:
As discussed, not needed.